README.md
# CNN Markdown Service

[![dependency-status](https://gemnasium.com/cnnlabs/cnn-markdown-service.svg)](https://gemnasium.com/cnnlabs/cnn-markdown-service)
[![build-status-master](https://img.shields.io/travis/cnnlabs/cnn-markdown-service/master.svg?style=flat-square&label=master)](https://travis-ci.org/cnnlabs/cnn-markdown-service)
[![build-status-develop](https://img.shields.io/travis/cnnlabs/cnn-markdown-service/master.svg?style=flat-square&label=develop)](https://travis-ci.org/cnnlabs/cnn-markdown-service)



## Requirements

Read these "_requirements_" as "_only tested with_".

- [Node.js](https://nodejs.org/) LTS or higher



## Install

```shell
$ npm install --save --save-exact cnn-markdown-service
```



## Usage

This is intended to be used as a dependency in a larger application.  Refer to
the [example.js](./example/example.js) that you can run with
`$ node example/example.js`.



## ESDoc Documentation

You can generate and view the docs locally with the commands below.  The `open`
command will only work on macOS.

```shell
$ npm run generate-docs
$ open docs/index.html
```



## NPM scripts

- `generate-authors` - Generates [AUTHORS.md](./AUTHORS.md).
- `generate-changelog` - Generates output to put in [CHANGELOG.md](./CHANGELOG.md).
- `generate-coverage` - Generates a code coverage report in `/coverage`.
- `generate-docs` - Generates ESDoc documentation in `/docs`.
- `test` - Runs all tests.
- `update-apply` - Updates [package.json](./package.json) with dependency updates.
- `update-check` - Outputs if any dependency updates are needed.



## Developer notes

- Always develop on the node version specified in the [.nvmrc](./.nvmrc) file.
  If [nvm](https://github.com/creationix/nvm) is used typing `nvm install`
  in the terminal will insure the correct version is used.

- Contributors should be familiar with the [Contributors Guide](https://github.com/cnnlabs/organization-docs/blob/master/CONTRIBUTING.md)

- Collaborators should be familiar with the [Collaborator Guide](https://github.com/cnnlabs/organization-docs/blob/master/COLLABORATOR_GUIDE.md)

- The current Project Owner (PO) of this project is Jamie Young ([@jamsyoung](https://github.com/jamsyoung/)).



## Licensing

See [LICENSE.md](./LICENSE.md) for details.


---
♥︎ 'Cause I'm A Turn It In And I'm A Turn It Out But Now I've Got To Pass The Mic To Yauch - King Ad-Rock


CHANGELOG.md
# CNN Markdown Service Changelog


## 2016-10-24, Version 1.1.0, @jamsyoung

### Notable changes

- Honor numbered lists, do not process them as markdown


### Known issues

See https://github.com/cnnlabs/cnn-markdown-service/labels/defect for complete and
current list of known issues.


# Commits

* [[`da6b2a3e71`](https://github.com/cnnlabs/cnn-markdown-service/commit/da6b2a3e71)] - **markdown-service**: add support for numbered lists (#5) (Katie Owen)
* [[`44038296c3`](https://github.com/cnnlabs/cnn-markdown-service/commit/44038296c3)] - **test**: add additional unit test around strong tag conversion (James Young)
* [[`d039e6237f`](https://github.com/cnnlabs/cnn-markdown-service/commit/d039e6237f)] - **test**: update testcase (James Young)



## 2016-07-07, Version 1.0.1, @jamsyoung

### Notable changes

- Fix a defect when strong tags included content with an asterisk


### Known issues

See https://github.com/cnnlabs/cnn-markdown-service/labels/defect for complete and
current list of known issues.


# Commits

* [[`c77a04e964`](https://github.com/cnnlabs/cnn-markdown-service/commit/c77a04e964)] - **license**: add licenses (James Young)
* [[`7f537d866b`](https://github.com/cnnlabs/cnn-markdown-service/commit/7f537d866b)] - **markdown-service**: properly handle <strong> tags that include asterisks (James Young)




## 2016-06-30, Version 1.0.0, @jamsyoung

### Notable changes

- Initial release


### Known issues

See https://github.com/cnnlabs/cnn-markdown-service/labels/defect for complete and
current list of known issues.


# Commits

All of them up to the v1.0.0 tag


test/mocha-setup.js
/*
 * Copyright 2016 Turner Broadcasting System, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const chai = require('chai');

chai.config.includeStack = false;

global.expect = chai.expect;
global.should = chai.should();
global.assert = chai.assert;

global.MarkdownService = require('../lib/markdown-service.js');


test/unit/markdown-service-test.js
/*
 * Copyright 2016 Turner Broadcasting System, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';


describe('Markdown Service', () => {
    beforeEach(() => {
        this.markdownService = new MarkdownService();
    });

    afterEach(() => {
        delete this.markdownService;
    });



    it('should instantiate properly when created properly', () => {
        this.markdownService.should.be.an.instanceof(MarkdownService);
    });

    it('should trim trailing spaces', () => {
        const response = this.markdownService.format('foo  '),
            expectedResponse = 'foo';

        response.should.equal(expectedResponse);
    });

    it('should trim leading spaces', () => {
        const response = this.markdownService.format('  foo'),
            expectedResponse = 'foo';

        response.should.equal(expectedResponse);
    });

    it('should trim trailing tabs', () => {
        const response = this.markdownService.format('foo\t\t'),
            expectedResponse = 'foo';

        response.should.equal(expectedResponse);
    });

    it('should trim leading tabs', () => {
        const response = this.markdownService.format('\t\tfoo'),
            expectedResponse = 'foo';

        response.should.equal(expectedResponse);
    });

    it('should trim trailing eol characters', () => {
        const response = this.markdownService.format('foo\n\n'),
            expectedResponse = 'foo';

        response.should.equal(expectedResponse);
    });

    it('should trim leading eol characters', () => {
        const response = this.markdownService.format('\n\nfoo'),
            expectedResponse = 'foo';

        response.should.equal(expectedResponse);
    });

    it('should convert two dashes (--) to an emdash (—) if they are surrounded by non-leading or trailing whitespace', () => {
        const response = this.markdownService.format('foo -- bar'),
            expectedResponse = 'foo — bar';

        response.should.equal(expectedResponse);
    });

    it('should convert multiple groups two dashes (--) to an emdash (—) if they are surrounded by non-leading or trailing whitespace and separated by something other than whitespace', () => {
        const response = this.markdownService.format('foo -- baz -- bar'),
            expectedResponse = 'foo — baz — bar';

        response.should.equal(expectedResponse);
    });

    it('should remove <em> tags if they surround only spaces', () => {
        const response = this.markdownService.format('<em>  </em>'),
            expectedResponse = '';

        response.should.equal(expectedResponse);
    });

    it('should remove <em> tags if they surround only tabs', () => {
        const response = this.markdownService.format('<em>\t\t</em>'),
            expectedResponse = '';

        response.should.equal(expectedResponse);
    });

    it('should remove <em> tags if they surround only new lines', () => {
        const response = this.markdownService.format('<em>\n\n</em>'),
            expectedResponse = '';

        response.should.equal(expectedResponse);
    });

    it('should remove <em> tags if they surround combinations of whitespace', () => {
        const response = this.markdownService.format('<em>  \t  \n \t</em>'),
            expectedResponse = '';

        response.should.equal(expectedResponse);
    });

    it('should remove <strong> tags if they surround only spaces', () => {
        const response = this.markdownService.format('<strong>  </strong>'),
            expectedResponse = '';

        response.should.equal(expectedResponse);
    });

    it('should remove <strong> tags if they surround only tabs', () => {
        const response = this.markdownService.format('<strong>\t\t</strong>'),
            expectedResponse = '';

        response.should.equal(expectedResponse);
    });

    it('should remove <strong> tags if they surround only new lines', () => {
        const response = this.markdownService.format('<strong>\n\n</strong>'),
            expectedResponse = '';

        response.should.equal(expectedResponse);
    });

    it('should remove <strong> tags if they surround combinations of whitespace', () => {
        const response = this.markdownService.format('<strong>  \t  \n \t</strong>'),
            expectedResponse = '';

        response.should.equal(expectedResponse);
    });

    it('should properly escape asterisks (*) that are intentional and properly work inside a <strong> tag', () => {
        const response = this.markdownService.format('<strong>*foo</strong>'),
            expectedResponse = '**\\*foo**';

        response.should.equal(expectedResponse);
    });

    it('should properly escape asterisks (*) that are intentional and properly work outside a <strong> tag', () => {
        const response = this.markdownService.format('*<strong>foo</strong>'),
            expectedResponse = '\\***foo**';

        response.should.equal(expectedResponse);
    });


    it('should remove <sub> tags since there is no markdown equivalent', () => {
        const response = this.markdownService.format('<sub>foo</sub>'),
            expectedResponse = 'foo';

        response.should.equal(expectedResponse);
    });

    it('should remove <u> tags since there is no markdown equivalent', () => {
        const response = this.markdownService.format('<u>foo</u>'),
            expectedResponse = 'foo';

        response.should.equal(expectedResponse);
    });

    it('should convert <br /> tags to a new line (\\n)', () => {
        const response = this.markdownService.format('foo<br />bar'),
            expectedResponse = 'foo\nbar';

        response.should.equal(expectedResponse);
    });

    it('should convert <br/> tags to a new line (\\n)', () => {
        const response = this.markdownService.format('foo<br/>bar'),
            expectedResponse = 'foo\nbar';

        response.should.equal(expectedResponse);
    });

    it('should convert <br> tags to a new line (\\n)', () => {
        const response = this.markdownService.format('foo<br>bar'),
            expectedResponse = 'foo\nbar';

        response.should.equal(expectedResponse);
    });

    it('should convert <em> tags to an underscore (_)', () => {
        const response = this.markdownService.format('<em>foo</em>'),
            expectedResponse = '_foo_';

        response.should.equal(expectedResponse);
    });

    it('should convert <strong> tags to two asterisks (**)', () => {
        const response = this.markdownService.format('<strong>foo</strong>'),
            expectedResponse = '**foo**';

        response.should.equal(expectedResponse);
    });

    it('should convert <a> tags with absolute urls to proper markdown anchor format', () => {
        const response = this.markdownService.format('<a href="http://www.cnn.com/">CNN</a>'),
            expectedResponse = '[CNN](http://www.cnn.com/)';

        response.should.equal(expectedResponse);
    });

    it('should convert <a> tags with relative urls to proper markdown anchor format', () => {
        const response = this.markdownService.format('<a href="/foo/bar">CNN</a>'),
            expectedResponse = '[CNN](/foo/bar)';

        response.should.equal(expectedResponse);
    });

    it('should convert <a> tags with properties other than href to proper markdown anchor format', () => {
        const response = this.markdownService.format('<a style="foo" href="http://www.cnn.com/" target="_blank">CNN</a>'),
            expectedResponse = '[CNN](http://www.cnn.com/)';

        response.should.equal(expectedResponse);
    });

    it('should convert multiple <a> tags with properties other than href to proper markdown anchor format in the same string', () => {
        const response = this.markdownService.format('<a style="foo" href="http://www.cnn.com/" target="_blank">CNN</a> <a style="bar" href="http://money.cnn.com/" target="_blank">CNN Money</a>'),
            expectedResponse = '[CNN](http://www.cnn.com/) [CNN Money](http://money.cnn.com/)';

        response.should.equal(expectedResponse);
    });

    it('should convert <h1> tags to prepend a hash (#) before the string', () => {
        const response = this.markdownService.format('<h1>Foo</h1>'),
            expectedResponse = '# Foo';

        response.should.equal(expectedResponse);
    });

    it('should convert <h2> tags to prepend two hashes (##) before the string', () => {
        const response = this.markdownService.format('<h2>Foo</h2>'),
            expectedResponse = '## Foo';

        response.should.equal(expectedResponse);
    });

    it('should convert <h3> tags to prepend three hashs (###) before the string', () => {
        const response = this.markdownService.format('<h3>Foo</h3>'),
            expectedResponse = '### Foo';

        response.should.equal(expectedResponse);
    });

    it('should convert <h4> tags to prepend four hashes (####) before the string', () => {
        const response = this.markdownService.format('<h4>Foo</h4>'),
            expectedResponse = '#### Foo';

        response.should.equal(expectedResponse);
    });

    it('should convert <h5> tags to prepend five hashes (#####) before the string', () => {
        const response = this.markdownService.format('<h5>Foo</h5>'),
            expectedResponse = '##### Foo';

        response.should.equal(expectedResponse);
    });

    it('should convert <h6> tags to prepend six hashes (######) before the string', () => {
        const response = this.markdownService.format('<h6>Foo</h6>'),
            expectedResponse = '###### Foo';

        response.should.equal(expectedResponse);
    });

    it('should convert single digit numbered lists to add \\\\ between the digit and the digit\'s "."', () => {
        const response = this.markdownService.format('1. Foo'),
            expectedResponse = '1\\\\. Foo';

        response.should.equal(expectedResponse);
    });

    it('should convert double digit numbered lists to add \\\\ between the digit and the digit\'s "."', () => {
        const response = this.markdownService.format('11. Foo'),
            expectedResponse = '11\\\\. Foo';

        response.should.equal(expectedResponse);
    });

    it('should return what was passed in if it is not a string, like an array', () => {
        const response = this.markdownService.format(['foo', 'bar']),
            expectedResponse = ['foo', 'bar'];

        response.should.be.an('array');
        response.should.deep.equal(expectedResponse);
    });

    it('should return what was passed in if it is not a string, like an object', () => {
        const response = this.markdownService.format({baz: ['foo', 'bar']}),
            expectedResponse = {baz: ['foo', 'bar']};

        response.should.be.an('object');
        response.should.deep.equal(expectedResponse);
    });
});


lib/markdown-service.js
/*
 * Copyright 2016 Turner Broadcasting System, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

/**
 * Formats HTML tags in strings as Markdown.  The intent is that each "string"
 * is a "paragraph".
 *
 * @example
 * const MarkdownService = require('cnn-markdown-service'),
 *     md = new MarkdownService(),
 *     paragraphs = [
 *         'I see it in the <a href="http://www.nytimes.com/2015/03/11/opinion/a-sensible-bill-on-medical-marijuana.html?_r=0" target="_blank">op-ed pages of the newspapers</a>, and on the state ballots in nearly half the country. I see it in politicians who once preferred to play it safe with this explosive issue but are now willing to stake their political futures on it. I see the revolution in the eyes of sterling scientists, previously reluctant to dip a toe into this heavily stigmatized world, who are diving in head first. I see it in the <a href="http://www.cnn.com/video/data/2.0/video/health/2015/04/16/gupta-murthy-medical-marijuana.cnn.html" target="_blank">new surgeon general</a> who cites data showing just how helpful it can be.',
 *         'I see a revolution in the attitudes of everyday Americans. For the first time a majority, <a href="http://www.people-press.org/2015/04/14/in-debate-over-legalizing-marijuana-disagreement-over-drugs-dangers/" target="_blank">53%, favor its legalization</a>, with <a href="http://www.people-press.org/2013/04/04/majority-now-supports-legalizing-marijuana/" target="_blank">77% supporting it for medical purposes</a>.'
 *     ];
 *
 * paragraphs.forEach((paragraph) => {
 *     console.log(md.format(paragraph));
 * });
 */
class MarkdownService {
    /**
     * Formats HTML tags in strings as Markdown.
     *
     * @param {String} data
     * A string with HTML tags that need to be formatted as Markdown.
     *
     * @returns {String}
     * Returns the string with HTML tags formatted as Markdown.
     */
    format(data) {
        if (typeof data === 'string') {
            // trim all leading and trailing whitespace
            data = data.trim();

            // convert two dashes to an emdash, if they are surrounded by whitespace
            data = data.replace(/(\s)--(\s)/g, '$1—$2');

            // remove known tags that do not wrap content, or only wrap whitespace
            data = data.replace(/<em>\s*<\/em>/g, '');
            data = data.replace(/<strong>\s*<\/strong>/g, '');

            // remove tags that have no markdown equivalent
            data = data.replace(/<sub>|<\/sub>/g, '');
            data = data.replace(/<u>|<\/u>/g, '');

            // convert HTML to Markdown
            data = data.replace(/<br\s*\/>/g, '\n');
            data = data.replace(/<br>/g, '\n');
            data = data.replace(/<em>|<\/em>/g, '_');
            data = data.replace(/\*/, '\\*');
            data = data.replace(/<\/li>/g, '');
            data = data.replace(/<strong>|<\/strong>/g, '**');

            // convert ol
            if (data.match(/<ol>/)) {
                data = data.replace(/<ol>|<\/ol>/g, '');
                data = data.replace(/<li>/g, '\n\n');
            }

            // convert ul
            if (data.match(/<ul>/)) {
                data = data.replace(/<ul>|<\/ul>/g, '');
                data = data.replace(/<li>/g, '\n\n- ');
            }

            if (data.match(/<a.*>|<\/a>/)) {
                // https://regex101.com/r/zQ7rD3/2
                data = data.replace(/<a.+?href="(.+?)".*?>(.+?)<\/a>/g, (match, p1, p2) => {
                    // encode spaces in urls
                    p1 = p1.replace(/\s+/g, '%20');

                    return `[${p2}](${p1})`;
                });
            }

            if (data.match(/^<h\d>.*<\/h\d>$/)) {
                data = `${'#'.repeat(data.match(/^<h(\d)>/)[1])} ${data.replace(/<h\d>|<\/h\d>/g, '')}`;
            }
        }

        return data;
    }
}



module.exports = MarkdownService;


example/example.js
/*
 * Copyright 2016 Turner Broadcasting System, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * Run this example with:
 *     $ node example/example.js
 */

'use strict';

const MarkdownService = require('../lib/markdown-service.js'),
    md = new MarkdownService(),
    paragraphs = [
        'I see signs of a revolution everywhere.',
        'I see it in the <a href="http://www.nytimes.com/2015/03/11/opinion/a-sensible-bill-on-medical-marijuana.html?_r=0" target="_blank">op-ed pages of the newspapers</a>, and on the state ballots in nearly half the country. I see it in politicians who once preferred to play it safe with this explosive issue but are now willing to stake their political futures on it. I see the revolution in the eyes of sterling scientists, previously reluctant to dip a toe into this heavily stigmatized world, who are diving in head first. I see it in the <a href="http://www.cnn.com/video/data/2.0/video/health/2015/04/16/gupta-murthy-medical-marijuana.cnn.html" target="_blank">new surgeon general</a> who cites data showing just how helpful it can be.',
        'I see a revolution in the attitudes of everyday Americans. For the first time a majority, <a href="http://www.people-press.org/2015/04/14/in-debate-over-legalizing-marijuana-disagreement-over-drugs-dangers/" target="_blank">53%, favor its legalization</a>, with <a href="http://www.people-press.org/2013/04/04/majority-now-supports-legalizing-marijuana/" target="_blank">77% supporting it for medical purposes</a>.',
        'Support for legalization has risen 11 points in the past few years alone. In 1969, the first time Pew asked the question about legalization, only 12% of the nation was in favor.',
        'I see a revolution that is burning white hot among young people, but also shows up among the parents and grandparents in my kids\' school. A police officer I met in Michigan is part of the revolution, as are the editors of the medical journal, <a href="http://journals.lww.com/neurosurgery/Fulltext/2015/04000/WEED___Marijuana,_Medicine_and_Neuroscience_.11.aspx#" target="_blank">Neurosurgery</a>. I see it in the faces of good parents, uprooting their lives to get medicine for their children -- and in the children themselves, <a href="http://www.cnn.com/2013/08/07/health/charlotte-child-medical-marijuana/">such as Charlotte</a>, who went from having 300 seizures a week to just one or two a month. We know it won\'t consistently have such dramatic results (or any impact at all) in others, but what medicine does?',
        'I see this medical marijuana revolution in surprising places.',
        '<a href="http://www.cnn.com/2015/04/16/us/georgia-medical-marijuana-bill-signing-haleigh-cox/index.html" target="_blank">Girl\'s seizures spur medical marijuana legislation in Georgia</a>',
        'Among my colleagues, my patients and my friends. I have even seen the revolution in my own family. A few years ago, when I told my mother I was investigating the topic for a documentary, I was met with a long pause.',
        '"Marijuana...?" She whispered in a half questioning, half disapproving tone. She could barely even say the word and her response filled me with self-doubt. Even as a grown man, mom can still make my cheeks turn red and shatter my confidence with a single word. But just last week she suddenly stopped mid-conversation and said, "I am proud of you on the whole marijuana thing." I waited for the other shoe to drop, but it didn\'t. Instead, she added, "You probably helped a lot of people who were suffering."',
        'I don\'t think we had ever had a conversation like that one. At that moment, I saw a revolution that can bring you to tears.',
        'The word revolution, comes from the Latin <em>revolutio</em>, to "turn around."',
        'I had <a href="http://www.cnn.com/2013/08/08/health/gupta-changed-mind-marijuana/">my own turn around</a> a couple of years ago, and at the time it was a lonely place to hold a supportive position on medical marijuana. Hardly any government officials would agree to sit down and be interviewed on the topic. Even patients I spoke to were reluctant to share their stories.',
        'It can be tricky, I learned, to be on the right side of science but on the wrong side of ideology.',
        'When we put the first "Weed" documentary on television in August 2013, I didn\'t know if anyone would watch our yearlong investigation. Even worse, I didn\'t even know if they would care.',
        '<a href="http://www.cnn.com/2015/01/07/us/recreational-marijuana-laws/index.html" target="_blank">Is weed legal in your state?</a>',
        'Just two years later, in "Weed 3," we are eyewitnesses to a revolution in full swing. You will ride along with us for the dawn of the first federally approved clinical study on the use of marijuana for PTSD. You will meet patients such as Sean Kiernan, an accomplished investment banker, and Amelia Taylor, a stay-at-home mom.',
        'They are the remarkable and surprising faces of this revolution -- smart, successful and suffering -- unwilling to accept the fact that commonly prescribed medications often used to treat PTSD can be worse than the underlying disorder itself. Sean Kiernan nearly died, trying to get better.',
        'You will see what weed really does to your brain, in crystal clear images. This time around, you will hear from the heads of government agencies earnestly sharing their point of view, both Democratic and Republican senators, and even the President of the United States.',
        'This is what a revolution looks like.',
        '<a href="http://www.cnn.com/2015/04/16/health/sanjay-gupta-medical-marijuana-questions-answers/index.html" target="_blank">Your medical marijuana questions answered</a>',
        'When "Weed 2: Cannabis Madness" aired in March 2014, Boston researcher Rick Doblin believed the right people were watching. Just four days later, Doblin received a letter in the mail he had been waiting on for seven years that finally provided federal approval for his marijuana study. The federal farm where Doblin would have to obtain his marijuana is on the campus of Ole Miss in Oxford, Mississippi. In anticipation of a scientific revolution, the production of research-grade marijuana there has increased 30-fold in just the past year.',
        'Make no mistake, we have plenty of evidence that the approval and support of the federal government can fast track a revolution at a faster pace than we have yet seen.',
        'It was the National Institute of Allergy and Infectious Diseases that spearheaded the research into a cure for AIDS, as well as stopping the spread of West Nile Virus. They were also responsible for the awesome task of eradicating polio and smallpox. Other successful federally backed programs include the human genome project, the BRAIN initiative and the Precision Medicine Initiative. There are no shortage of examples where the federal government has been a guardian of our public health needs, and you could argue that medical marijuana would also qualify as a worthwhile investment.',
        '<a href="http://www.cnn.com/2015/04/15/health/marijuana-medical-advances/index.html">10 diseases where medical marijuana could have impact</a>',
        'There is now promising research into the use of marijuana that could impact tens of thousands of children and adults, including treatment for cancer, epilepsy and Alzheimer\'s, to name a few. With regard to pain alone, marijuana could greatly reduce the demand for narcotics and simultaneously decrease the number of accidental painkiller overdoses, which are the greatest cause of preventable death in this country.',
        'As I sat across from Sens. Kirsten Gillibrand (D-New York) and Cory Booker (D-New Jersey), I knew something extraordinary was happening.',
        'They were reciting the story of Charlotte Figi and countless other children. They were quoting back the data we had shared from our earlier investigations. They were extolling the potential virtues of the plant, and all of that was before the interview even started. There was an impatience about them, and they seemed in a hurry to make a large dent in marijuana reform.',
        'They want marijuana to be rescheduled. They want it now.',
        'They want doctors to be able to prescribe it at VA hospitals all over the country. They want it now.',
        'They want research dollars freed up to study the plant. They want it now.',
        'They want their fellow lawmakers at the state and national level to acknowledge what most of the world, including the citizens of the United States, have known for a long time: Marijuana is a medicine, that should be studied and treated like any other medicine.',
        'And they want all of it now.',
        'I spent much of our interview challenging them. I needed to remind them that people, long before me or them, have been trying to do many of these same things for 40 years, and had been rejected every time. I reminded them that politicians have a hard time winning elections on the issue of marijuana but less difficulty losing them. I challenged them every step of the way. "This time will be different," Booker confidently told me as he walked out of the room.',
        '<a href="http://www.cnn.com/2014/01/20/health/marijuana-versus-alcohol/index.html" target="_blank">Is marijuana as safe as -- or safer than -- alcohol?</a>',
        'I know how easy it is do nothing because I did nothing for too long. Take a good look at the data, educate yourself and talk to the patients, who are often out of options and find their hope in the form of a simple plant.',
        'Journalists shouldn\'t take a position. It makes sense. Objectivity is king. But, at some point, open questions do get answered. At some point, contentious issues do get resolved. At some point, common sense prevails.',
        'So, here it is: We should legalize medical marijuana. We should do it nationally. And, we should do it now.',
        '<a href="http://money.cnn.com/2015/04/13/news/pot-marijuana-facts/" target="_blank">9 things to know about legal pot</a>',
        '2. "Hillary Clinton as you know, as most people know, she\'s a world class liar."',
        '12. "Hillary Clinton as you know, as most people know, she\'s a world class liar."'
    ];

paragraphs.forEach((paragraph) => {
    console.log(md.format(paragraph));
});
